File: typescript/systemdesign/bloomFilter.ts
/*
Bloom filter: probabilistic set membership over a fixed bit array using
multiple independent hash functions (derived via double hashing from two
base hashes). False positives are possible, false negatives never are.

status - completed
*/

import { logp } from "../utils/logger";

function fnv1a(value: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < value.length; i++) {
    hash ^= value.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

function djb2(value: string): number {
  let hash = 5381;
  for (let i = 0; i < value.length; i++) {
    hash = (hash * 33) ^ value.charCodeAt(i);
  }
  return hash >>> 0;
}

class BloomFilter {
  private readonly bits: Uint8Array;
  private readonly sizeBits: number;

  constructor(sizeBits: number, private readonly numHashes: number = 4) {
    this.sizeBits = sizeBits;
    this.bits = new Uint8Array(sizeBits);
  }

  // double hashing: combine two base hashes to cheaply simulate `numHashes`
  // independent hash functions, per Kirsch-Mitzenmacher
  private indexesFor(value: string): number[] {
    const h1 = fnv1a(value);
    const h2 = djb2(value);
    const indexes: number[] = [];
    for (let i = 0; i < this.numHashes; i++) {
      const combined = (h1 + i * h2) >>> 0;
      indexes.push(combined % this.sizeBits);
    }
    return indexes;
  }

  add(value: string): void {
    for (const idx of this.indexesFor(value)) {
      this.bits[idx] = 1;
    }
  }

  mightContain(value: string): boolean {
    return this.indexesFor(value).every((idx) => this.bits[idx] === 1);
  }
}

// usage scenarios

const filter = new BloomFilter(1000, 4);
const added = ["alice", "bob", "carol", "dave"];
added.forEach((name) => filter.add(name));

for (const name of added) {
  logp(`mightContain("${name}") -> ${filter.mightContain(name)} (expected true, was added)`);
}

const notAdded = ["eve", "mallory", "trent"];
let falsePositives = 0;
for (const name of notAdded) {
  const result = filter.mightContain(name);
  if (result) falsePositives++;
  logp(`mightContain("${name}") -> ${result} (not added; true would be a false positive)`);
}
logp(`false positives among ${notAdded.length} unadded names: ${falsePositives}`);

// demonstrate false-positive rate stays low with a well-sized filter and a
// larger batch of never-added values
const largeFilter = new BloomFilter(10_000, 4);
for (let i = 0; i < 500; i++) {
  largeFilter.add(`present-${i}`);
}
let fpCount = 0;
const trials = 1000;
for (let i = 0; i < trials; i++) {
  if (largeFilter.mightContain(`absent-${i}`)) fpCount++;
}
logp(`false positive rate over ${trials} unadded keys: ${((100 * fpCount) / trials).toFixed(2)}% (expected low, well under 100%)`);

File: python/systemdesign/bloom_filter.py
"""
Bloom filter: probabilistic set membership backed by a bit array and
multiple independent hash functions (via double hashing). Membership
checks can false-positive but never false-negative, and the filter
uses far less memory than storing the actual set.

status - completed
"""

import hashlib

from base_logger.logging_event import create_logger


class BloomFilter:

    log = create_logger(__name__)

    def __init__(self, size: int = 1000, num_hashes: int = 5):
        self.size = size
        self.num_hashes = num_hashes
        self.bits = [0] * size

    def _hashes(self, item: str):
        item_bytes = item.encode("utf-8")
        h1 = int(hashlib.md5(item_bytes).hexdigest(), 16)
        h2 = int(hashlib.sha1(item_bytes).hexdigest(), 16)
        for i in range(self.num_hashes):
            yield (h1 + i * h2) % self.size

    def add(self, item: str):
        for index in self._hashes(item):
            self.bits[index] = 1

    def might_contain(self, item: str) -> bool:
        return all(self.bits[index] == 1 for index in self._hashes(item))


if __name__ == "__main__":
    log = create_logger("bloom_filter_demo")

    bloom = BloomFilter(size=2000, num_hashes=5)

    added = [f"user-{i}@example.com" for i in range(200)]
    for email in added:
        bloom.add(email)

    all_added_found = all(bloom.might_contain(email) for email in added)
    log.info("all 200 added items report present: %s (expected True, no false negatives)",
             all_added_found)

    not_added = [f"stranger-{i}@example.com" for i in range(500)]
    false_positives = sum(1 for email in not_added if bloom.might_contain(email))
    log.info("false positives among 500 never-added items: %d / %d (%.1f%%)",
             false_positives, len(not_added), 100.0 * false_positives / len(not_added))
    log.info("expected a small but nonzero false positive rate, never a false negative")
